tests/governance_js.py

```diff
@@ -298,6 +313,77 @@ def test_pure_proposals(network, args):
     return network
 
 
+@reqs.description("Test proposal replay protection")
+def test_proposal_replay_protection(network, args):
+    node = network.find_random_node()
+
+    with node.client(None, None, "member0") as c:
+        # Creating a proposal with too large a created_at always fails
+        c.set_created_at_override(int("1" + "0" * 10))
+        r = c.post("/gov/proposals", always_accept_noop)
+        assert (
+            r.status_code == 400
+            and r.body.json()["error"]["code"] == "InvalidCreatedAt"
+        ), r.body.text()
+
+        # Fill window size with proposals
+        window_size = 100
+        now = int(datetime.now().timestamp()) - 500
+        submitted = []
+        for i in range(window_size):
+            c.set_created_at_override(now + i)
+            proposal = unique_always_accept_noop()
+            r = c.post("/gov/proposals", proposal)
+            assert r.status_code == 200, r.body.text()
+            submitted.append(proposal)
+
+        # Re-submitting the last proposal is detected as a replay
+        last_index = window_size - 1
+        c.set_created_at_override(now + last_index)
+        r = c.post("/gov/proposals", submitted[last_index])
+        assert (
+            r.status_code == 400 and r.body.json()["error"]["code"] == "ProposalReplay"
+        ), r.body.text()
+
+        # Submitting proposals earlier than, or in the first half of the window is rejected
+        c.set_created_at_override(now - 1)
+        r = c.post("/gov/proposals", always_accept_noop)
+        assert (
+            r.status_code == 400
+            and r.body.json()["error"]["code"] == "ProposalCreatedTooLongAgo"
+        ), r.body.text()
+
+        c.set_created_at_override(now + (window_size // 2) - 1)
+        r = c.post("/gov/proposals", always_accept_noop)
+        assert (
+            r.status_code == 400
+            and r.body.json()["error"]["code"] == "ProposalCreatedTooLongAgo"
+        ), r.body.text()
+
+        # Submitting a unique proposal just past the median of the window does work
+        c.set_created_at_override(now + (window_size // 2))
+        r = c.post("/gov/proposals", unique_always_accept_noop())
+        assert r.status_code == 200, r.body.text()
+
+        r = c.post("/gov/proposals", set_service_recent_cose_proposals_window_size(1))
+        assert r.status_code == 200, r.body.text()
```

Comment:
Why does this proposal not have to go through the usual propose/vote flow?

Reply:
Proximally, because of https://github.com/microsoft/CCF/pull/4823/files#diff-ba4676ee17cf56956ba8162d50f0046f14347b7a1be803618365a20b2f4b744c. There isn't meaningful voting in this case, and it's not what we're testing, so like much of the rest of the scenario, we skip it.